File: 080_MaxPathSum.py
"""
Problem: Maximum Path Sum
Description: Find the maximum path sum in a binary tree. Path can start
             and end at any node.

Approach: At each node, calculate max path through it and update global max.
"""

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def maxPathSum(root: Optional[TreeNode]) -> int:
    max_sum = [float('-inf')]
    
    def helper(node):
        if not node:
            return 0
        
        left = max(0, helper(node.left))
        right = max(0, helper(node.right))
        
        max_sum[0] = max(max_sum[0], left + right + node.val)
        
        return node.val + max(left, right)
    
    helper(root)
    return max_sum[0]

def main():
    root = TreeNode(-10)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)
    
    print(maxPathSum(root))  # Output: 42

if __name__ == "__main__":
    main()

"""
Time Complexity: O(n)
Space Complexity: O(h)
"""

File: 080_MaxPathSum.cpp
/*
Problem: Maximum Path Sum
Description: Find the maximum path sum in a binary tree. Path can start
             and end at any node.

Approach: At each node, calculate max path through it and update global max.
*/

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left, *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

int maxPathHelper(TreeNode* root, int& maxSum) {
    if (!root) return 0;
    
    int left = max(0, maxPathHelper(root->left, maxSum));
    int right = max(0, maxPathHelper(root->right, maxSum));
    
    maxSum = max(maxSum, left + right + root->val);
    
    return root->val + max(left, right);
}

int maxPathSum(TreeNode* root) {
    int maxSum = INT_MIN;
    maxPathHelper(root, maxSum);
    return maxSum;
}

int main() {
    TreeNode* root = new TreeNode(-10);
    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);
    
    cout << maxPathSum(root) << endl;  // Output: 42
    
    return 0;
}

/*
Time Complexity: O(n)
Space Complexity: O(h)
*/
